File: graph_reachability_using_dfs/grudfs.py
def can_reach(arr: list[int], start: int) -> bool:
    n = len(arr)
    stack = [start]
    visited: set[int] = set()

    while stack:
        i = stack.pop()

        if i < 0 or i >= n or i in visited:
            continue

        if arr[i] == 0:
            return True

        visited.add(i)

        stack.append(i + arr[i])
        stack.append(i - arr[i])

    return False


print(can_reach([4, 2, 3, 0, 3, 1, 2], 5))
print(can_reach([4, 2, 3, 0, 3, 1, 2], 0))
print(can_reach([3, 0, 2, 1, 2], 2))

File: graph_reachability_using_dfs/grudfs.ts
function canReach(arr: number[], start: number): boolean {
    const n = arr.length;
    const stack: number[] = [start];
    const visited = new Set<number>();

    while (stack.length > 0) {
        const i = stack.pop() as number;

        if (i < 0 || i >= n || visited.has(i)) {
            continue;
        }

        if (arr[i] === 0) {
            return true;
        }

        visited.add(i);

        stack.push(i + arr[i]);
        stack.push(i - arr[i]);
    }

    return false;
}

console.log(canReach([4, 2, 3, 0, 3, 1, 2], 5));
console.log(canReach([4, 2, 3, 0, 3, 1, 2], 0));
console.log(canReach([3, 0, 2, 1, 2], 2));
